File: leetcode/medium/InsertIntoABinarySearchTree.py
from typing import Optional


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        """
        Inserts a value into a Binary Search Tree (BST).

        This solution uses a recursive approach, which is a natural fit for tree
        traversals and manipulations. The logic leverages the core property of a BST:
        - Values in the left subtree are smaller than the root's value.
        - Values in the right subtree are larger than the root's value.

        Algorithm:
        1.  Base Case: If the current node (`root`) is None, we have found the correct
            position to insert the new value. We create a new `TreeNode` with `val`
            and return it.
        2.  Recursive Step:
            - If `val` is greater than the current node's value, it belongs in the
              right subtree. We recursively call `insertIntoBST` on the right child
              (`root.right`) and assign the result back to `root.right`.
            - If `val` is less than the current node's value, it belongs in the left
              subtree. We do the same for the left child (`root.left`).
        3.  Return the (potentially modified) `root` of the current subtree.

        Time Complexity: O(H), where H is the height of the tree. O(log N) for a balanced
        tree, O(N) for a skewed tree.
        Space Complexity: O(H) due to the recursion stack.
        """
        # Base case: If we've reached an empty spot, insert the new node here.
        if not root:
            return TreeNode(val)

        # Recursive step: Traverse down the tree to find the correct position.
        if val > root.val:
            root.right = self.insertIntoBST(root.right, val)
        else:  # val < root.val
            root.left = self.insertIntoBST(root.left, val)

        return root

File: leetcode/medium/InsertIntoABinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    /**
     * Inserts a new value into a Binary Search Tree (BST) while maintaining the BST property.
     *
     * This implementation uses an iterative approach to find the insertion location, 
     * which optimizes space complexity by avoiding the recursion call stack overhead.
     *
     * Time Complexity: O(h), where h is the height of the tree.
     * - O(log n) for a balanced tree, O(n) for a skewed (worst-case) tree.
     * Space Complexity: O(1), as only a constant number of pointers are used (no recursion stack).
     *
     * @param root The root of the BST. Can be null if the tree is empty.
     * @param val The value to be inserted into the BST.
     * @return The root of the BST after insertion.
     */
    public TreeNode insertIntoBST(TreeNode root, int val) {
        // If the tree is empty, create a new node and return it as the root.
        if (root == null) {
            return new TreeNode(val);
        }

        // Initialize pointers for traversal.
        TreeNode current = root;
        TreeNode parent = null;

        // Traverse the tree to find the correct leaf node location.
        while (current != null) {
            parent = current;
            
            // Move left if the new value is smaller.
            if (val < current.val) {
                current = current.left;
            } 
            // Move right if the new value is larger or equal (BST standard).
            else {
                current = current.right;
            }
        }

        // 'parent' is now the node to which the new node must be attached.
        TreeNode newNode = new TreeNode(val);
        
        // Determine whether to attach the new node to the left or right of 'parent'.
        if (val < parent.val) {
            parent.left = newNode; // Attach as the left child.
        } else {
            parent.right = newNode; // Attach as the right child.
        }

        // Return the original root.
        return root;
    }
}